File: packages/protocol/src/codec.ts
const HEADER_SIZE = 8;

/**
 * Maximum allowed frame payload (jsonLen + binLen) in bytes.
 *
 * Rationale: the largest legitimate frames we send today are 1 MiB binary
 * PTY sidecars (codec-edge.test.ts) plus their JSON metadata (a few hundred
 * bytes). The QueuedWriter caps its in-memory queue at 8 MiB. A 64 MiB ceiling
 * is therefore 64× the real-world peak — generous enough to absorb any future
 * growth, but tight enough to reject a poison header declaring 4 GiB
 * (0xFFFFFFFF binLen) before we attempt to buffer anything. Frames exceeding
 * this limit indicate a protocol violation (corrupt peer, adversarial input,
 * or a stream sync loss) and require connection teardown.
 */
const MAX_FRAME_SIZE = 64 * 1024 * 1024; // 64 MiB

const encoder = new TextEncoder();
const decoder = new TextDecoder();

/**
 * A decoded frame. `binary` is the raw payload that accompanies the JSON when
 * the sender used the binary-sidecar path (e.g. Runner → Daemon PTY io); it
 * is `null` for plain JSON frames.
 */
export interface DecodedFrame {
  data: unknown;
  binary: Uint8Array<ArrayBufferLike> | null;
}

/**
 * Wire format v2:
 *   u32_be jsonLen
 *   u32_be binLen  (0 for plain JSON frames)
 *   utf-8 JSON
 *   binLen bytes of raw binary payload (only when binLen > 0)
 *
 * The binary-sidecar path lets high-throughput streams (PTY io records) avoid
 * base64-encoding the payload into the JSON body — a ~33% size saving plus
 * one fewer copy on each side. Receivers read the JSON first (for routing
 * metadata) and then treat `binary` as the actual payload.
 */
export function encodeFrame(
  data: unknown,
  binary?: Uint8Array<ArrayBufferLike> | null,
): Uint8Array {
  const json = encoder.encode(JSON.stringify(data));
  const binLen = binary?.byteLength ?? 0;
  const frame = new Uint8Array(HEADER_SIZE + json.byteLength + binLen);
  const view = new DataView(frame.buffer);
  view.setUint32(0, json.byteLength);
  view.setUint32(4, binLen);
  frame.set(json, HEADER_SIZE);
  if (binary && binLen > 0) {
    frame.set(binary, HEADER_SIZE + json.byteLength);
  }
  return frame;
}

export class FrameDecoder {
  // ArrayBufferLike (vs the stricter default ArrayBuffer) so that incoming
  // chunks backed by SharedArrayBuffer or Buffer are assignable without a
  // copy — the decode path only reads, it never resizes or transfers.
  private buf: Uint8Array<ArrayBufferLike> = new Uint8Array(0);

  /**
   * Consume a chunk of bytes. Returns every complete frame the chunk (plus
   * any carried-over tail) contains. Each frame is `{ data, binary }` where
   * `binary` is `null` unless the sender attached a binary sidecar.
   */
  decode(chunk: Uint8Array<ArrayBufferLike>): DecodedFrame[] {
    // Common case: the previous chunk left nothing pending. Skip the concat
    // entirely — `chunk` is sufficient as-is. For a 1 MiB burst split into N
    // chunks this avoids O(N²) bytes of buffer growth.
    if (this.buf.byteLength === 0) {
      this.buf = chunk;
    } else {
      const next = new Uint8Array(this.buf.byteLength + chunk.byteLength);
      next.set(this.buf);
      next.set(chunk, this.buf.byteLength);
      this.buf = next;
    }

    const results: DecodedFrame[] = [];

    while (this.buf.byteLength >= HEADER_SIZE) {
      const view = new DataView(
        this.buf.buffer,
        this.buf.byteOffset,
        this.buf.byteLength,
      );
      const jsonLen = view.getUint32(0);
      const binLen = view.getUint32(4);
      const totalLen = HEADER_SIZE + jsonLen + binLen;

      // H1: Reject oversized frames immediately so a poison header (e.g.
      // binLen=0xFFFFFFFF) cannot cause unbounded buffer growth / OOM.
      // This is an unrecoverable protocol violation — throw so the caller
      // tears down the connection rather than leaving the decoder wedged.
      if (jsonLen + binLen > MAX_FRAME_SIZE) {
        throw new Error(
          `Frame too large: ${jsonLen + binLen} bytes exceeds max ${MAX_FRAME_SIZE}`,
        );
      }

      if (this.buf.byteLength < totalLen) break;

      const json = decoder.decode(
        this.buf.subarray(HEADER_SIZE, HEADER_SIZE + jsonLen),
      );
      let binary: Uint8Array<ArrayBufferLike> | null = null;
      if (binLen > 0) {
        // Detach the binary slice so the caller can safely hold onto it past
        // the next decode() — otherwise a later `new Uint8Array(this.buf)`
        // would only copy the tail, leaving this view aliased to a chunk
        // buffer the caller owns.
        const start = HEADER_SIZE + jsonLen;
        binary = new Uint8Array(
          this.buf.subarray(start, start + binLen).slice().buffer,
        );
      }

      // M1: JSON.parse throws on malformed/empty JSON — advance buf past the
      // frame first so the decoder is not permanently wedged on re-entry, then
      // re-throw so the caller tears down the connection (consistent with H1).
      this.buf = this.buf.subarray(totalLen);
      results.push({ data: JSON.parse(json), binary });
    }

    // Detach the tail from `chunk` when it's only a partial frame, so the
    // caller's chunk buffer can be GC'd and so the next `decode` doesn't
    // alias into caller-owned memory.
    if (this.buf.byteLength > 0 && this.buf.buffer === chunk.buffer) {
      this.buf = new Uint8Array(this.buf);
    }

    return results;
  }

  reset(): void {
    this.buf = new Uint8Array(0);
  }
}

File: rust/tp-core/src/codec.rs
//! Framed-JSON codec — byte-exact port of `packages/protocol/src/codec.ts`.
//!
//! Wire format v2:
//! ```text
//!   u32_be jsonLen
//!   u32_be binLen   (0 for plain JSON frames)
//!   utf-8 JSON
//!   binLen bytes of raw binary payload (only when binLen > 0)
//! ```
//!
//! `HEADER_SIZE = 8`, `MAX_FRAME_SIZE = 64 MiB`. The binary-sidecar path lets
//! high-throughput streams (PTY io) avoid base64-encoding the payload.

use crate::error::{Result, TpError};

pub const HEADER_SIZE: usize = 8;
/// 64 MiB — matches codec.ts. Rejects poison headers (e.g. binLen=0xFFFFFFFF)
/// before any allocation.
pub const MAX_FRAME_SIZE: usize = 64 * 1024 * 1024;

/// A decoded frame. `binary` is `None` for plain JSON frames; `Some` when the
/// sender attached a binary sidecar (binLen > 0).
#[derive(Debug, Clone, PartialEq)]
pub struct DecodedFrame {
    /// The JSON payload bytes (UTF-8), exactly as they appeared on the wire.
    pub json: Vec<u8>,
    pub binary: Option<Vec<u8>>,
}

/// Encode `json` (raw UTF-8 JSON bytes) plus optional `binary` sidecar into a
/// single framed buffer. Byte-identical to `encodeFrame(JSON.parse(json), binary)`
/// in codec.ts (the caller is responsible for producing the JSON exactly as the
/// TS `JSON.stringify` would — see `serialize_envelope`).
pub fn encode_frame(json: &[u8], binary: Option<&[u8]>) -> Vec<u8> {
    let bin_len = binary.map(|b| b.len()).unwrap_or(0);
    let mut frame = Vec::with_capacity(HEADER_SIZE + json.len() + bin_len);
    frame.extend_from_slice(&(json.len() as u32).to_be_bytes());
    frame.extend_from_slice(&(bin_len as u32).to_be_bytes());
    frame.extend_from_slice(json);
    if let Some(b) = binary {
        if bin_len > 0 {
            frame.extend_from_slice(b);
        }
    }
    frame
}

/// Incremental frame decoder mirroring `FrameDecoder` in codec.ts. Feed it
/// arbitrary byte chunks; it returns every complete frame contained in the
/// accumulated buffer and carries over any partial tail.
#[derive(Default)]
pub struct FrameDecoder {
    buf: Vec<u8>,
}

impl FrameDecoder {
    pub fn new() -> Self {
        Self { buf: Vec::new() }
    }

    /// Consume `chunk`, returning every complete frame now available. Errors
    /// (oversized frame, malformed JSON length) are unrecoverable protocol
    /// violations — the buffer is advanced past the offending frame where
    /// possible, but the caller should tear down the connection.
    pub fn decode(&mut self, chunk: &[u8]) -> Result<Vec<DecodedFrame>> {
        self.buf.extend_from_slice(chunk);
        let mut results = Vec::new();

        let mut offset = 0usize;
        while self.buf.len() - offset >= HEADER_SIZE {
            let json_len = u32::from_be_bytes(
                self.buf[offset..offset + 4].try_into().unwrap(),
            ) as usize;
            let bin_len = u32::from_be_bytes(
                self.buf[offset + 4..offset + 8].try_into().unwrap(),
            ) as usize;

            // H1: reject oversized frames before attempting to buffer the body.
            if json_len + bin_len > MAX_FRAME_SIZE {
                return Err(TpError::Frame(format!(
                    "Frame too large: {} bytes exceeds max {}",
                    json_len + bin_len,
                    MAX_FRAME_SIZE
                )));
            }

            let total_len = HEADER_SIZE + json_len + bin_len;
            if self.buf.len() - offset < total_len {
                break;
            }

            let json_start = offset + HEADER_SIZE;
            let json = self.buf[json_start..json_start + json_len].to_vec();
            let binary = if bin_len > 0 {
                let bin_start = json_start + json_len;
                Some(self.buf[bin_start..bin_start + bin_len].to_vec())
            } else {
                None
            };

            offset += total_len;
            results.push(DecodedFrame { json, binary });
        }

        // Drop consumed bytes from the front, keep any partial tail.
        if offset > 0 {
            self.buf.drain(0..offset);
        }
        Ok(results)
    }

    pub fn reset(&mut self) {
        self.buf.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encode_json_only_header() {
        let json = br#"{"t":"x"}"#;
        let frame = encode_frame(json, None);
        assert_eq!(&frame[0..4], &(json.len() as u32).to_be_bytes());
        assert_eq!(&frame[4..8], &0u32.to_be_bytes());
        assert_eq!(&frame[8..], json);
    }

    #[test]
    fn round_trip_with_binary() {
        let json = br#"{"k":"io"}"#;
        let bin = [1u8, 2, 3, 4, 5];
        let frame = encode_frame(json, Some(&bin));
        let mut dec = FrameDecoder::new();
        let frames = dec.decode(&frame).unwrap();
        assert_eq!(frames.len(), 1);
        assert_eq!(frames[0].json, json);
        assert_eq!(frames[0].binary.as_deref(), Some(&bin[..]));
    }

    #[test]
    fn split_chunks_reassemble() {
        let json = br#"{"hi":1}"#;
        let frame = encode_frame(json, None);
        let mut dec = FrameDecoder::new();
        assert!(dec.decode(&frame[0..3]).unwrap().is_empty());
        assert!(dec.decode(&frame[3..6]).unwrap().is_empty());
        let frames = dec.decode(&frame[6..]).unwrap();
        assert_eq!(frames.len(), 1);
        assert_eq!(frames[0].json, json);
    }

    #[test]
    fn two_frames_in_one_chunk() {
        let mut buf = encode_frame(br#"{"a":1}"#, None);
        buf.extend_from_slice(&encode_frame(br#"{"b":2}"#, None));
        let mut dec = FrameDecoder::new();
        let frames = dec.decode(&buf).unwrap();
        assert_eq!(frames.len(), 2);
        assert_eq!(frames[0].json, br#"{"a":1}"#);
        assert_eq!(frames[1].json, br#"{"b":2}"#);
    }

    #[test]
    fn rejects_oversized_frame() {
        // jsonLen = 0xFFFFFFFF poison header.
        let mut poison = vec![0xFFu8, 0xFF, 0xFF, 0xFF, 0, 0, 0, 0];
        poison.extend_from_slice(b"x");
        let mut dec = FrameDecoder::new();
        assert!(matches!(dec.decode(&poison), Err(TpError::Frame(_))));
    }
}
